Tests for EVP KEM API (positive and negative tests).

bhess opened the pull request:
Adds tests for EVP KEM API (positive and negative tests).

Adds a negative tests for EVP Signature API.

Fix in hybrid KEM registration -> don't register X448 hybrids with 256 bit security (as already done with the TLS groups).

One thing I noticed while running the negative KEM tests (with modified ciphertext): HQC is the only KEM that returns an error code when decapsulation fails. The other KEMs only allow an implicit check if decapsulation was successful by comparing the original with the decapsulated message, while the return code is always OQS_SUCCESS. This is mainly to prevent leaking information.
As with the negative KEM tests in liboqs, the tests here let both behavior pass. But we may want to document this behavior or add an extra test for it.

Fixes #57

baentsch reviewed (changes requested):
Basically LGTM. Indeed, the "X-algs" were not properly matched. Thanks also for the additional negative sig test. Would it be reasonable to add an explanation (explicit list?) to the README regarding the  hybrids supported by `oqsprovider` (as for L1 & L3 there's now still the additional X... hybrids that are not supported/mentioned at https://github.com/open-quantum-safe/openssl#supported-algorithms that we otherwise link to)? 

Finally, I'd definitely suggest adding wording/warning that the algorithms (except HQC as you state) effectively don't support the OpenSSL API (not returning an error code when they failed) -- I'd be fine not mentioning this if there'd be wider agreement that "silent failure" is an acceptable security feature, e.g., by adding an explicit statement to this effect to https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_decapsulate.html) .

bhess commented:
> Would it be reasonable to add an explanation (explicit list?) to the README regarding the hybrids supported by `oqsprovider` (as for L1 & L3 there's now still the additional X... hybrids that are not supported/mentioned at https://github.com/open-quantum-safe/openssl#supported-algorithms that we otherwise link to)?

I added a description to the README about the additional x25519/x448 hybrids, similar to the description in OQS-OpenSSL.



> Finally, I'd definitely suggest adding wording/warning that the algorithms (except HQC as you state) effectively don't support the OpenSSL API (not returning an error code when they failed) -- I'd be fine not mentioning this if there'd be wider agreement that "silent failure" is an acceptable security feature, e.g., by adding an explicit statement to this effect to https://www.openssl.org/docs/manmaster/man3/EVP_PKEY_decapsulate.html) .

Yes I agree. For now I added a note to the README. Will open an issue in OpenSSL about the specification of the decaps API.

baentsch reviewed (approved):
Thx!
